### DISCUSSION TITLE
github packages maven integration

### DISCUSSION THREAD

**Discussion by tombrus** [community/community#26528] [Actions]
<p>I am looking into <em>github packages</em> as a replacement for publishing our libraries on <em>maven-central</em>. </p>
<ol>
<li>is it valid to expect that <em>github packages</em> is a replacement for <em>maven-central</em>?</li>
</ol>
<p>We expect that other projects that use our open source libraries should be able to specify a maven dependency in their <strong>pom</strong> and add a github repo in their <strong>settings</strong>.</p>
<ol start="2">
<li>is this a valid assumption?</li>
</ol>
<p>When both questions above get a ‘<em>yes</em>’, we have a number of questions:</p>
<ol start="3">
<li>
<p>I would expect that projects would add only one github repo where all published artifacts can be found. But the repo url seems to have the name of our <strong>organisation</strong> and our <strong>github project</strong> in it which makes it specific for our artifacts. Should projects add a github repo in their settings file for <strong><em>every</em></strong> library supplier separately? This does not seem to add up…</p>
</li>
<li>
<p>On maven central you do not need to authenticate to <strong>get</strong> artifacts but on github packages you always need to authenticate, even for reading. This means that my CI needs a valid login or token just to read dependencies. Although perfectly possibble to implement this seems like an extra hurdle. What is the reasoning behind this?</p>
</li>
<li>
<p>Does (or will) gihub packages support the <a href="https://repository.sonatype.org/nexus-restlet1x-plugin/default/docs/rest.html" rel="nofollow noopener">nexus REST API</a>? We tend to access artifacts through this interface from shell scripts.</p>
</li>
</ol>
<p>I hope someone can give us some answers…</p>
<p>Tom Brus</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Sorry for late response, please check the answers below, thanks.</p>
<ol>
<li>Yes, you can choose to use Package Registry as a replacement for Maven Central. However Package Registry currently is more suited for ‘private’ packages because of the requirement to authenticate(like below point3).</li>
<li>You can create a single repo to hold all of your packages, or create one repo per package or any combination therein. It‘s entirely up to the organization/user on how they prefer to work.</li>
<li>Correct, you currently are required to authenticate using either a <a href="https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line" rel="nofollow noopener">Personal Access Token</a>(PAT) or if you are using GitHub Actions for CI, you can use the provided token as mentioned <a href="https://help.github.com/en/github/managing-packages-with-github-packages/using-github-packages-with-github-actions" rel="nofollow noopener">here</a>.</li>
<li>You can manage packages with GitHub API, please refer to the <a href="https://help.github.com/en/github/managing-packages-with-github-packages/about-github-packages#managing-packages-with-the-github-api" rel="nofollow noopener">link</a> for more details, thanks.</li>
</ol>

**Comment by bdelion** [comment]
<p>Hello <a class="mention" href="/weide-zhou">@weide-zhou</a> ,</p>
<p>For point 2., how should it be done?<br>
I tried to do this for a Java Maven project, but it doesn’t work:</p>
<ul>
<li>a project for the pom parent</li>
<li>a project for the Java application which is based on the pom parent</li>
<li>a project to store all my packages.<br>
Are there examples or tutorials?</li>
</ul>

**Comment by bdelion** [comment]
<p>Hello,</p>
<p>It’s finally work for me with the good declaration in settings.xml and PAT.</p>

**Comment by gnahraf** [comment]
I'm trying to understand the _value_ in publishing maven packages from a user's perspective, and honestly, I see little value compared to publishing to a central repo. Having to add repo URLs per organization in the user's pom does not scale. Until Github makes its own central maven repo, I see little value in publishing packages this way.

PS I think there's opportunity in making a GitHub Maven Central's credential process way smoother (since GitHub can already authenticate its users).
